src/qtfirebasedb.h

```diff
@@ -0,0 +1,199 @@
+        kErrorWriteCanceled = firebase::database::kErrorWriteCanceled,
+        kErrorInvalidVariantType = firebase::database::kErrorInvalidVariantType,
+        kErrorConflictingOperationInProgress = firebase::database::kErrorConflictingOperationInProgress,
+        kErrorTransactionAbortedByUser = firebase::database::kErrorTransactionAbortedByUser,
+    };
+    Q_ENUM(Error)
+
+    //TODO implement database related functions
+    //(not important to write/read data for now but useful for extended control)
+private:
+    explicit QtFirebaseDb(QObject *parent = 0);
+    void init() override;
+    void onFutureEvent(QString eventId, firebase::FutureBase future) override;
+
+    QString getFutureKey(QtFirebaseDbRequest* request) const;
+    QtFirebaseDbRequest* getRequest(const QString& futureKey) const;
+    void addFuture(QString requestId, QtFirebaseDbRequest* request, firebase::FutureBase future);
+    void unregisterRequest(QtFirebaseDbRequest* request);
+    QString prefix() const;
+private:
+    static QtFirebaseDb* self;
+    Q_DISABLE_COPY(QtFirebaseDb)
+
+    firebase::database::Database* m_db;
+    QMap<QString, QtFirebaseDbRequest*> m_requests;
+    QMutex m_futureMutex;
+
+    friend class QtFirebaseDbRequest;
+};
+
+class QtFirebaseDataSnapshot: public QObject
+{
+    Q_OBJECT
+public:
+    QtFirebaseDataSnapshot(const firebase::database::DataSnapshot& snapshot);
+public slots:
+    bool exists() const;
+    QString key() const;
+    QVariant value() const;
+    QByteArray jsonString() const;
+    bool hasChildren() const;
+    bool isValid() const;
+
+    //not implemented firebase functions
+    //
+    //currently can be accessed through qt json support (except priority), like
+    //QJsonDocument doc = QJsonDocument::fromVariant(request->getSnapshot().value());
+    //
+
+    /*QtFirebaseDataSnapshot* child(const QString& path);
+    /*size_t childrenCount() const;
+    QtFirebaseDataSnapshot getChild(int i);
+    QVariant priority() const;
+    bool hasChild(const QString& path) const;
+    */
+private:
+    const firebase::database::DataSnapshot& m_snapshot;
+};
+
+class QtFirebaseDbRequest;
+class QtFirebaseDbQuery: public QObject
+{
+    Q_OBJECT
+public:
+    QtFirebaseDbQuery();
+public slots:
+    QtFirebaseDbQuery* orderByKey();
+    QtFirebaseDbQuery* orderByValue();
+    QtFirebaseDbQuery* orderByChild(const QString& path);
+    QtFirebaseDbQuery* orderByPriority();
+    QtFirebaseDbQuery* startAt(QVariant order_value);
+    QtFirebaseDbQuery* startAt(QVariant order_value, const QString& child_key);
+    QtFirebaseDbQuery* endAt(QVariant order_value);
+    QtFirebaseDbQuery* endAt(QVariant order_value, const QString& child_key);
+    QtFirebaseDbQuery* equalTo(QVariant order_value);
+    QtFirebaseDbQuery* equalTo(QVariant order_value, const QString& child_key);
+    QtFirebaseDbQuery* limitToFirst(size_t limit);
+    QtFirebaseDbQuery* limitToLast(size_t limit);
+    void getValue();
```

Comment:
I think we should omit `get` here as well:
getValue -> value

Reply:
done